packages/webservice/app/src/main/kotlin/com/yahoo/navi/ws/utilities/CarbonQueryLogger.kt

```diff
@@ -0,0 +1,153 @@
+import java.util.Timer
+import java.util.TimerTask
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import javax.persistence.EntityManager
+import javax.persistence.PersistenceContext
+import javax.transaction.Transactional
+import javax.ws.rs.core.MultivaluedMap
+
+object Constants {
+    const val MAX_QUERIES = 5
+    const val TIMEOUT = 10000
+}
+
+@Transactional
+open class CarbonQueryLogger constructor(
+    maxQueries: Int = Constants.MAX_QUERIES,
+    timeOut: Long = Constants.TIMEOUT.toLong()
+) : QueryLogger, TimerTask() {
+
+    private var queriesToBeLogged: MutableList<QueryStats> = ArrayList()
+    private var queryWatcher: MutableMap<UUID, QueryStats> = ConcurrentHashMap()
+
+    @PersistenceContext
+    private lateinit var entityManager: EntityManager
+
+    private var t: Timer = Timer()
+    private val maxQueries: Int = maxQueries
+    private val timeOut: Long = timeOut
+
+    init {
+        t.schedule(this, 0, this.timeOut)
+    }
+
+    override fun cancelQuery(queryId: UUID) {
+        var endTime: Long = System.currentTimeMillis()
+        var qstat: QueryStats = queryWatcher[queryId]!!
+        synchronized(this) {
+            qstat.status = QueryStatus.CANCELLED
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = 0
+            qstat.rowsReturned = 0
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
+    }
+
+    override fun acceptQuery(
+        queryId: UUID,
+        user: User?,
+        headers: MutableMap<String, String>?,
+        apiVer: String?,
+        queryParams: Optional<MultivaluedMap<String, String>>,
+        path: String?
+    ) {
+        var start: Long = System.currentTimeMillis()
+        var apiQuery: String? = constructAPIQuery(queryParams, path)
+        var userName: String? = user?.name
+
+        var qStat = QueryStats(queryId)
+        qStat.user = userName
+        qStat.apiVersion = apiVer
+        qStat.apiQuery = apiQuery
+        qStat.startTime = start
+        qStat.status = QueryStatus.ACCEPTED
+        qStat.hostName = getLocalHost().hostName
+        queryWatcher[queryId] = qStat
+    }
+
+    override fun completeQuery(queryId: UUID, response: QueryResponse) {
+        var endTime: Long = System.currentTimeMillis()
+        var qstat: QueryStats = queryWatcher[queryId]!!
+        var status: QueryStatus = if (response.data == null) QueryStatus.FAILED else QueryStatus.COMPLETED
+        synchronized(this) {
+            qstat.status = status
+            qstat.durationInMillis = endTime - qstat.startTime!!
+            qstat.bytesReturned = getTotalBytesReturned(response)
+            qstat.rowsReturned = getTotalRowsReturned(response)
+            queriesToBeLogged.add(qstat)
+            queryWatcher.remove(queryId)
+        }
+        if (queriesToBeLogged.size > maxQueries) { run() }
```

Comment:
Fetch queriesToBeLogged.size() inside the synchronized block.  YOu can use the value outside as you do here.

Reply:
Alternatively, move the entire condition inside the synchronized block.